### PR TITLE
EXEC-2603 - Allow disabling rate limit backoff

### PR DESCRIPTION
**Pull request by DomParfitt**
Adds an opt-in to disable the existing backoff behaviour when receiving a `429` from the server.

### CODE REVIEW

**Review by joshhsoj1902 — APPROVED** [review]

File: httpclient/httpclient_test.go [hunk 1]
```diff
@@ -645,3 +645,53 @@ func TestClient_GETWithBody(t *testing.T) {
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := httpclient.New(httpclient.Config{
+				BaseURL:                 server.URL,
+				Timeout:                 5 * time.Second,
+				DisableRateLimitBackoff: tt.disableBackoff,
+			})
+			err := client.Call(ctx, httpclient.NewRequest("POST", "/",
+				httpclient.Timeout(time.Millisecond),
```
Comment (pete-woods) [inline httpclient/httpclient_test.go:676]: This timeout seems too short

File: httpclient/httpclient_test.go [hunk 2]
```diff
@@ -645,3 +645,53 @@ func TestClient_GETWithBody(t *testing.T) {
...
+			))
+			// confirm we get an http 429 error on first call
+			assert.Check(t, httpclient.HasStatusCode(err, http.StatusTooManyRequests), err)
+			// confirm that it is now not a warning
+			assert.Check(t, !o11y.IsWarning(err))
+
+			err = client.Call(ctx, httpclient.NewRequest("POST", "/",
+				httpclient.Timeout(time.Millisecond),
```
Comment (pete-woods) [inline httpclient/httpclient_test.go:684]: Same here

File: httpclient/httpclient.go [hunk 3]
```diff
@@ -62,19 +62,25 @@ type Config struct {
 	Tracer tracer
 	// DialContext allows a dial context to be injected into the HTTP transport.
 	DialContext func(ctx context.Context, network string, addr string) (net.Conn, error)
+	// DisableRateLimitBackoff disables the circuit breaker behaviour when the client receives a 429
+	// response. This is useful if the server rate-limits more granularly than requests-per-client
+	// (e.g. if it uses specific query params or headers when bucketing requests) but should be enabled
+	// with care as it can lead to thrashing the server if not used appropriately.
+	DisableRateLimitBackoff bool
```
Comment (pete-woods) [inline httpclient/httpclient.go:69]: The other place we have a "disable" feature it's called `NoRetry`, perhaps it makes sense to name this similarly `NoRateLimitBackoff`?

Comment (pete-woods) [reply httpclient/httpclient.go:69]: https://github.com/circleci/ex/blob/d3725f732b2dfe0df3a23d59721a5aa2851b8cc4/httpclient/httpclient.go#L323

**Review by danmux — APPROVED** [review]

**Review by pete-woods — APPROVED** [review]

**Pull request merged**